read-fonts/src/tables/colr.rs

```diff
@@ -1,5 +1,696 @@
+    /// may be used to detect recursion in the paint graph.
+    pub fn v1_base_glyph(
+        &self,
+        glyph_id: GlyphId,
+    ) -> Result<Option<(Paint<'a>, PaintId)>, ReadError> {
+        let list = self.base_glyph_list().ok_or(ReadError::NullOffset)??;
+        let records = list.base_glyph_paint_records();
+        let record = match records.binary_search_by(|rec| rec.glyph_id().cmp(&glyph_id)) {
+            Ok(ix) => &records[ix],
+            _ => return Ok(None),
+        };
+        let offset_data = list.offset_data();
+        // Use the address of the paint as an identifier for the recursion
+        // blacklist.
+        let id = record.paint_offset().to_u32() as usize + offset_data.as_ref().as_ptr() as usize;
+        Ok(Some((record.paint(offset_data)?, id)))
+    }
+
+    /// Returns the COLRv1 layer at the given index.
+    ///
+    /// The second value in the tuple is a unique identifier for the paint that
+    /// may be used to detect recursion in the paint graph.
+    pub fn v1_layer(&self, index: usize) -> Result<(Paint<'a>, PaintId), ReadError> {
+        let list = self.layer_list().ok_or(ReadError::NullOffset)??;
+        let offset = list
+            .paint_offsets()
+            .get(index)
+            .ok_or(ReadError::OutOfBounds)?
+            .get();
+        let offset_data = list.offset_data();
+        // Use the address of the paint as an identifier for the recursion
+        // blacklist.
+        let id = offset.to_u32() as usize + offset_data.as_ref().as_ptr() as usize;
+        Ok((offset.resolve(offset_data)?, id))
+    }
+
+    /// Returns the COLRv1 clip box for the given glyph identifier.
+    pub fn v1_clip_box(&self, glyph_id: GlyphId) -> Result<Option<ClipBox<'a>>, ReadError> {
+        use std::cmp::Ordering;
+        let list = self.clip_list().ok_or(ReadError::NullOffset)??;
+        let clips = list.clips();
+        let clip = match clips.binary_search_by(|clip| {
+            if glyph_id < clip.start_glyph_id() {
+                Ordering::Greater
+            } else if glyph_id > clip.end_glyph_id() {
+                Ordering::Less
+            } else {
+                Ordering::Equal
+            }
+        }) {
+            Ok(ix) => &clips[ix],
+            _ => return Ok(None),
+        };
+        Ok(Some(clip.clip_box(list.offset_data())?))
+    }
+}
+
+/// Combination of a `COLR` table and a location in variation space for
+/// resolving paints.
+///
+/// See [`Paint::resolve`], [`ColorStops::resolve`] and [`ClipBox::resolve`].
+pub struct ColrInstance<'a> {
+    colr: Colr<'a>,
+    index_map: Option<DeltaSetIndexMap<'a>>,
+    var_store: Option<ItemVariationStore<'a>>,
+    coords: &'a [F2Dot14],
+}
+
+impl<'a> ColrInstance<'a> {
+    /// Creates a new resolver for the given `COLR` table and normalized variation
+    /// coordinates.
+    pub fn new(colr: Colr<'a>, coords: &'a [F2Dot14]) -> Self {
+        let index_map = colr.var_index_map().and_then(|res| res.ok());
```

Comment:
just trivia but you can also write this as `colr.var_index_map().transpose().ok().flatten()`.

Reply:
I _think_ I had the transpose version in there first and clippy yelled at me about it 🤷‍♂️